Should support pattern matching guards

<!--
BEFORE SUBMITTING AN ISSUE:
1.  Search for your issue on GitHub: https://github.com/jhipster/prettier-java/issues
    A large number of opened issues are duplicates of existing issues.
    If someone has already opened an issue for what you are experiencing, you do not need to open a new issue — please add a 👍 reaction to the existing issue instead.
2.  We get a lot of requests for adding options, but Prettier is
    built on the principle of being opinionated about code formatting.
    This means we have a very high bar for adding new options.
    Find out more: https://prettier.io/docs/en/option-philosophy.html
-->

**Prettier-Java 1.6.1**

**Input:**

```java
package com.vimat.model;

public record Buyer(String name, double bestPrice, double joker) {
  public boolean hasBestOffer(Buyer other) {
      return switch (other) { 
        case null -> true;
        case Buyer b && this.bestPrice > b.bestPrice -> true;
        default -> false;
      };
    }
}
```

**Output:**

```
[ERROR] src/main/java/com/vimat/model/Buyer.java: Error: Sad sad panda, parsing errors detected in line: 7, column: 20!
[ERROR] Expecting --> '->' <-- but found --> '&&' <--!
[ERROR]         ->compilationUnit
[ERROR]         ->ordinaryCompilationUnit
[ERROR]         ->typeDeclaration
[ERROR]         ->classDeclaration
[ERROR]         ->recordDeclaration
[ERROR]         ->recordBody
[ERROR]         ->recordBodyDeclaration
[ERROR]         ->classBodyDeclaration
[ERROR]         ->classMemberDeclaration
[ERROR]         ->methodDeclaration
[ERROR]         ->methodBody
[ERROR]         ->block
[ERROR]         ->blockStatements
[ERROR]         ->blockStatement
[ERROR]         ->statement
[ERROR]         ->statementWithoutTrailingSubstatement
[ERROR]         ->returnStatement
[ERROR]         ->expression
[ERROR]         ->ternaryExpression
[ERROR]         ->binaryExpression
[ERROR]         ->unaryExpression
[ERROR]         ->primary
[ERROR]         ->primaryPrefix
[ERROR]         ->switchStatement
[ERROR]         ->switchBlock
[ERROR]         ->switchRule
[ERROR]     at Object.parse (/home/vinz/.m2/repository/com/hubspot/maven/plugins/prettier-maven-plugin/0.17/prettier-maven-plugin-0.17-prettier-java-1.6.1/prettier-java/node_modules/java-parser/src/index.js:41:11)
[ERROR]     at Object.parse (/home/vinz/.m2/repository/com/hubspot/maven/plugins/prettier-maven-plugin/0.17/prettier-maven-plugin-0.17-prettier-java-1.6.1/prettier-java/node_modules/prettier-plugin-java/dist/parser.js:4:26)
[ERROR]     at Object.parse$a [as parse] (/home/vinz/.m2/repository/com/hubspot/maven/plugins/prettier-maven-plugin/0.17/prettier-maven-plugin-0.17-prettier-java-1.6.1/prettier-java/node_modules/prettier/index.js:13652:19)
[ERROR]     at coreFormat (/home/vinz/.m2/repository/com/hubspot/maven/plugins/prettier-maven-plugin/0.17/prettier-maven-plugin-0.17-prettier-java-1.6.1/prettier-java/node_modules/prettier/index.js:15183:16)
[ERROR]     at formatWithCursor$1 (/home/vinz/.m2/repository/com/hubspot/maven/plugins/prettier-maven-plugin/0.17/prettier-maven-plugin-0.17-prettier-java-1.6.1/prettier-java/node_modules/prettier/index.js:15423:14)
[ERROR]     at Object.formatWithCursor (/home/vinz/.m2/repository/com/hubspot/maven/plugins/prettier-maven-plugin/0.17/prettier-maven-plugin-0.17-prettier-java-1.6.1/prettier-java/node_modules/prettier/index.js:60171:12)
[ERROR]     at format$1 (/home/vinz/.m2/repository/com/hubspot/maven/plugins/prettier-maven-plugin/0.17/prettier-maven-plugin-0.17-prettier-java-1.6.1/prettier-java/node_modules/prettier/bin-prettier.js:16640:21)
[ERROR]     at Object.formatFiles$1 [as formatFiles] (/home/vinz/.m2/repository/com/hubspot/maven/plugins/prettier-maven-plugin/0.17/prettier-maven-plugin-0.17-prettier-java-1.6.1/prettier-java/node_modules/prettier/bin-prettier.js:16754:22)
[ERROR]     at async main (/home/vinz/.m2/repository/com/hubspot/maven/plugins/prettier-maven-plugin/0.17/prettier-maven-plugin-0.17-prettier-java-1.6.1/prettier-java/node_modules/prettier/bin-prettier.js:18616:5)
[ERROR]     at async Object.run (/home/vinz/.m2/repository/com/hubspot/maven/plugins/prettier-maven-plugin/0.17/prettier-maven-plugin-0.17-prettier-java-1.6.1/prettier-java/node_modules/prettier/bin-prettier.js:18559:5)
```

**Expected behavior:**

Should support pattern matching guards

--- a/packages/java-parser/api.d.ts
+++ b/packages/java-parser/api.d.ts
@@ -375,6 +375,7 @@ export abstract class JavaCstVisitor<IN, OUT> implements ICstVisitor<IN, OUT> {
   arrayAccessSuffix(ctx: ArrayAccessSuffixCtx, param?: IN): OUT;
   methodReferenceSuffix(ctx: MethodReferenceSuffixCtx, param?: IN): OUT;
   pattern(ctx: PatternCtx, param?: IN): OUT;
+  primaryPattern(ctx: PrimaryPatternCtx, param?: IN): OUT;
   typePattern(ctx: TypePatternCtx, param?: IN): OUT;
   identifyNewExpressionType(ctx: IdentifyNewExpressionTypeCtx, param?: IN): OUT;
   isLambdaExpression(ctx: IsLambdaExpressionCtx, param?: IN): OUT;
@@ -625,6 +626,8 @@ export abstract class JavaCstVisitorWithDefaults<IN, OUT>
   switchBlock(ctx: SwitchBlockCtx, param?: IN): OUT;
   switchBlockStatementGroup(ctx: SwitchBlockStatementGroupCtx, param?: IN): OUT;
   switchLabel(ctx: SwitchLabelCtx, param?: IN): OUT;
+  caseOrDefaultLabel(ctx: CaseOrDefaultLabelCtx, param?: IN): OUT;
+  caseLabelElement(ctx: CaseLabelElementCtx, param?: IN): OUT;
   switchRule(ctx: SwitchRuleCtx, param?: IN): OUT;
   caseConstant(ctx: CaseConstantCtx, param?: IN): OUT;
   whileStatement(ctx: WhileStatementCtx, param?: IN): OUT;
@@ -728,6 +731,7 @@ export abstract class JavaCstVisitorWithDefaults<IN, OUT>
   arrayAccessSuffix(ctx: ArrayAccessSuffixCtx, param?: IN): OUT;
   methodReferenceSuffix(ctx: MethodReferenceSuffixCtx, param?: IN): OUT;
   pattern(ctx: PatternCtx, param?: IN): OUT;
+  primaryPattern(ctx: PrimaryPatternCtx, param?: IN): OUT;
   typePattern(ctx: TypePatternCtx, param?: IN): OUT;
   identifyNewExpressionType(ctx: IdentifyNewExpressionTypeCtx, param?: IN): OUT;
   isLambdaExpression(ctx: IsLambdaExpressionCtx, param?: IN): OUT;
@@ -1848,6 +1852,8 @@ export type IdentifyClassBodyDeclarationTypeCtx = {
   Volatile?: IToken[];
   Synchronized?: IToken[];
   Native?: IToken[];
+  Sealed?: IToken[];
+  NonSealed?: IToken[];
   Strictfp?: IToken[];
   unannType: UnannTypeCstNode[];
 };
@@ -2370,11 +2376,13 @@ export type IdentifyInterfaceBodyDeclarationTypeCtx = {
   Public?: IToken[];
   Protected?: IToken[];
   Private?: IToken[];
-  Static?: IToken[];
-  Final?: IToken[];
   Abstract?: IToken[];
-  Default?: IToken[];
+  Static?: IToken[];
+  Sealed?: IToken[];
+  NonSealed?: IToken[];
   Strictfp?: IToken[];
+  Final?: IToken[];
+  Default?: IToken[];
   unannType: UnannTypeCstNode[];
 };
 
@@ -2621,7 +2629,7 @@ export interface SwitchBlockStatementGroupCstNode extends CstNode {
 export type SwitchBlockStatementGroupCtx = {
   switchLabel: SwitchLabelCstNode[];
   Colon: IToken[];
-  blockStatements: BlockStatementsCstNode[];
+  blockStatements?: BlockStatementsCstNode[];
 };
 
 export interface SwitchLabelCstNode extends CstNode {
@@ -2652,10 +2660,10 @@ export interface CaseLabelElementCstNode extends CstNode {
 }
 
 export type CaseLabelElementCtx = {
-  caseConstant?: CaseConstantCstNode[];
-  pattern?: PatternCstNode[];
   Null?: IToken[];
   Default?: IToken[];
+  pattern?: PatternCstNode[];
+  caseConstant?: CaseConstantCstNode[];
 };
 
 export interface SwitchRuleCstNode extends CstNode {
@@ -3489,7 +3497,21 @@ export interface PatternCstNode extends CstNode {
 }
 
 export type PatternCtx = {
-  typePattern: TypePatternCstNode[];
+  primaryPattern: PrimaryPatternCstNode[];
+  AndAnd?: IToken[];
+  binaryExpression?: BinaryExpressionCstNode[];
+};
+
+export interface PrimaryPatternCstNode extends CstNode {
+  name: "primaryPattern";
+  children: PrimaryPatternCtx;
+}
+
+export type PrimaryPatternCtx = {
+  LBrace?: IToken[];
+  pattern?: PatternCstNode[];
+  RBrace?: IToken[];
+  typePattern?: TypePatternCstNode[];
 };
 
 export interface TypePatternCstNode extends CstNode {

--- a/packages/java-parser/src/productions/expressions.js
+++ b/packages/java-parser/src/productions/expressions.js
@@ -562,7 +562,26 @@ function defineRules($, t) {
   });
 
   $.RULE("pattern", () => {
-    $.SUBRULE($.typePattern);
+    $.SUBRULE($.primaryPattern);
+    $.OPTION(() => {
+      $.CONSUME(t.AndAnd);
+      $.SUBRULE($.binaryExpression);
+    });
+  });
+
+  $.RULE("primaryPattern", () => {
+    $.OR([
+      {
+        ALT: () => {
+          $.CONSUME(t.LBrace);
+          $.SUBRULE($.pattern);
+          $.CONSUME(t.RBrace);
+        }
+      },
+      {
+        ALT: () => $.SUBRULE($.typePattern)
+      }
+    ]);
   });
 
   $.RULE("typePattern", () => {

--- a/packages/java-parser/test/pattern-matching/pattern-matching-spec.js
+++ b/packages/java-parser/test/pattern-matching/pattern-matching-spec.js
@@ -32,4 +32,21 @@ describe("Pattern matching", () => {
     `;
     expect(() => javaParser.parse(input, "methodDeclaration")).to.not.throw();
   });
+
+  it("should support pattern matching guards", () => {
+    const input = `
+    package com.vimat.model;
+
+    public record Buyer(String name, double bestPrice, double joker) {
+      public boolean hasBestOffer(Buyer other) {
+          return switch (other) {
+            case null -> true;
+            case Buyer b && this.bestPrice > b.bestPrice -> true;
+            default -> false;
+          };
+        }
+    }
+    `;
+    expect(() => javaParser.parse(input, "compilationUnit")).to.not.throw();
+  });
 });

--- a/packages/prettier-plugin-java/src/options.js
+++ b/packages/prettier-plugin-java/src/options.js
@@ -172,6 +172,7 @@ module.exports = {
       { value: "arrayAccessSuffix" },
       { value: "methodReferenceSuffix" },
       { value: "pattern" },
+      { value: "primaryPattern" },
       { value: "typePattern" },
       { value: "identifyNewExpressionType" },
       { value: "isLambdaExpression" },

--- a/packages/prettier-plugin-java/src/printers/expressions.ts
+++ b/packages/prettier-plugin-java/src/printers/expressions.ts
@@ -34,6 +34,7 @@ import {
   ParenthesisExpressionCtx,
   PatternCtx,
   PrimaryCtx,
+  PrimaryPatternCtx,
   PrimaryPrefixCtx,
   PrimarySuffixCtx,
   PrimitiveCastExpressionCtx,
@@ -75,6 +76,7 @@ import {
   sortAnnotationIdentifier,
   sortNodes
 } from "./printer-utils";
+import join = builders.join;
 
 const { ifBreak, line, softline, indentIfBreak } = builders;
 
@@ -759,7 +761,28 @@ export class ExpressionsPrettierVisitor extends BaseCstPrettierPrinter {
   }
 
   pattern(ctx: PatternCtx) {
-    return this.visitSingle(ctx);
+    const primaryPattern = this.visit(ctx.primaryPattern);
+    if (ctx.AndAnd === undefined) {
+      return primaryPattern;
+    }
+
+    const binaryExpression = this.visit(ctx.binaryExpression);
+    return rejectAndConcat([
+      primaryPattern,
+      " ",
+      ctx.AndAnd[0],
+      line,
+      binaryExpression
+    ]);
+  }
+
+  primaryPattern(ctx: PrimaryPatternCtx) {
+    if (ctx.LBrace === undefined) {
+      return this.visitSingle(ctx);
+    }
+
+    const pattern = this.visit(ctx.pattern);
+    return putIntoBraces(pattern, softline, ctx.LBrace[0], ctx.RBrace![0]);
   }
 
   typePattern(ctx: TypePatternCtx) {

--- a/packages/prettier-plugin-java/test/unit-test/pattern-matching/_input.java
+++ b/packages/prettier-plugin-java/test/unit-test/pattern-matching/_input.java
@@ -12,4 +12,24 @@ class T {
         }
         return formatted;
     }
+
+    void test(Buyer other) {
+        return switch (other) {
+            case null -> true;
+            case Buyer b && this.bestPrice > b.bestPrice -> true;
+            case Buyer b && this.bestPrice > b.bestPrice -> {
+                return true;
+            }
+            case (Buyer b && this.bestPrice > b.bestPrice) -> true;
+            case Buyer b && this.bestPrice > b.bestPrice && this.bestPrice > b.bestPrice && this.bestPrice > b.bestPrice && this.bestPrice > b.bestPrice -> true;
+            case Buyer b && this.bestPrice > b.bestPrice && this.bestPrice > b.bestPrice && this.bestPrice > b.bestPrice && this.bestPrice > b.bestPrice -> {
+                return true;
+            }
+            case (Buyer b && this.bestPrice > b.bestPrice && this.bestPrice > b.bestPrice && this.bestPrice > b.bestPrice && this.bestPrice > b.bestPrice) -> true;
+            case (Buyer b && this.bestPrice > b.bestPrice && this.bestPrice > b.bestPrice && this.bestPrice > b.bestPrice && this.bestPrice > b.bestPrice) -> {
+                return true;
+            }
+            default -> false;
+        };
+    }
 }

--- a/packages/prettier-plugin-java/test/unit-test/pattern-matching/_output.java
+++ b/packages/prettier-plugin-java/test/unit-test/pattern-matching/_output.java
@@ -18,4 +18,44 @@ class T {
     }
     return formatted;
   }
+
+  void test(Buyer other) {
+    return switch (other) {
+      case null -> true;
+      case Buyer b && this.bestPrice > b.bestPrice -> true;
+      case Buyer b && this.bestPrice > b.bestPrice -> {
+        return true;
+      }
+      case (Buyer b && this.bestPrice > b.bestPrice) -> true;
+      case Buyer b &&
+        this.bestPrice > b.bestPrice &&
+        this.bestPrice > b.bestPrice &&
+        this.bestPrice > b.bestPrice &&
+        this.bestPrice > b.bestPrice -> true;
+      case Buyer b &&
+        this.bestPrice > b.bestPrice &&
+        this.bestPrice > b.bestPrice &&
+        this.bestPrice > b.bestPrice &&
+        this.bestPrice > b.bestPrice -> {
+        return true;
+      }
+      case (
+          Buyer b &&
+          this.bestPrice > b.bestPrice &&
+          this.bestPrice > b.bestPrice &&
+          this.bestPrice > b.bestPrice &&
+          this.bestPrice > b.bestPrice
+        ) -> true;
+      case (
+          Buyer b &&
+          this.bestPrice > b.bestPrice &&
+          this.bestPrice > b.bestPrice &&
+          this.bestPrice > b.bestPrice &&
+          this.bestPrice > b.bestPrice
+        ) -> {
+        return true;
+      }
+      default -> false;
+    };
+  }
 }

